File: 1014-best-sightseeing-pair/1014-best-sightseeing-pair.java
class Solution {
    public int maxScoreSightseeingPair(int[] values) {
        int[] values1 = values.clone();
        int result = Integer.MIN_VALUE;
        for (int i = 1; i < values1.length; i++) {
            values1[i] = Math.max(values1[i-1], values[i] + i);
            result = Math.max(result, values1[i-1] + values[i] - i);
        }
        return result;
    }
}

File: 1014-best-sightseeing-pair/1014-best-sightseeing-pair.py
class Solution:
    def maxScoreSightseeingPair(self, values: List[int]) -> int:
        max_ = float('-inf')
        dp = values.copy()
        for i in range(1, len(dp)):
            dp[i] = max(dp[i-1], dp[i]+i)
            max_ = max(max_, values[i] - i + dp[i-1])
        return max_